kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -401,35 +408,77 @@ private static void validateSchemaEvolution(
+    for (Map.Entry<Integer, StructFieldWithIds> entry : newFields.entrySet()) {
+      StructFieldWithIds existing = currentFields.get(entry.getKey());
+      // For top level columns, their parents will be null so that this equivalence check
+      // generalizes
+      if (existing != null && !Objects.equals(existing.parentId(), entry.getValue().parentId())) {
+        throw new KernelException(
+            String.format(
+                "Cannot move field %s outside of its containing struct", existing.columnPath()));
+      }
+    }
+  }
+
   /**
    * Verifies that no non-nullable fields are added, no existing field nullability is tightened and
    * no invalid type changes are performed
-   *
-   * <p>ToDo: Prevent moving fields outside of their containing struct
    */
-  private static void validateUpdatedSchemaCompatibility(SchemaChanges schemaChanges) {
-    for (StructField addedField : schemaChanges.addedFields()) {
-      if (!addedField.isNullable()) {
+  private static void validateUpdatedSchemaCompatibility(
+      SchemaChanges<StructFieldWithIds> schemaChanges,
+      Map<Integer, StructFieldWithIds> currentFields,
+      boolean icebergCompatV2) {
+    for (StructFieldWithIds addedField : schemaChanges.addedFields()) {
+      if (!addedField.field().isNullable()) {
         throw new KernelException(
-            String.format("Cannot add non-nullable field %s", addedField.getName()));
+            String.format("Cannot add non-nullable field %s", addedField.field().getName()));
       }
     }
 
-    for (Tuple2<StructField, StructField> updatedFields : schemaChanges.updatedFields()) {
+    for (Tuple2<StructFieldWithIds, StructFieldWithIds> updatedFields :
+        schemaChanges.updatedFields()) {
+      // In IcebergCompatV2, map keys where keys are structs cannot be modified beyond field renames
+      if (icebergCompatV2) {
+        StructFieldWithIds updatedField = updatedFields._2;
+        if (updatedField.parentId() != null) {
+          DataType parentDataType =
+              currentFields.get(updatedField.parentId()).field().getDataType();
+          // Check if a map key where the key is a struct has been modified
+          // It's safe here to use columnPath to determine if the field is the map key instead of
+          // the value because
+          // we know that the parent is a map type and we know that the path internally assigned to
+          // a map
+          // key includes "key"
+          if (updatedField.field().getDataType() instanceof StructType
+              && parentDataType instanceof MapType
+              && updatedField.columnPath().contains("key")) {
+            throw new KernelException(
+                "Cannot modify map field %s with struct key in icebergCompatV2");
+          }
+        }
+      }
+
       // ToDo: See if recursion can be avoided by incorporating map key/value and array element
```

Comment:
After this change in the case of IcebergCompatV2 we can actually avoid recursion entirely because now the fields include map.key/value array element. In the non IcebergCompatV2 case we don't have those fields registered as changes since they don't have IDs, but we still need to do validations on the map key/v or array element fields don't go through invalid changes. 

To make an optimization generalize for non icebergCompatV2 we may need to somehow have a sentinel ID for map key/value or array element that indicates that it's not a real ID but more so just a placeholder to make the optimized diff computation generalize across both cases. I'll table this for now